### DISCUSSION TITLE
Shading Ice Shelf with Polartoolkit

### DISCUSSION THREAD

**Discussion by YingpuRin** [mdtanker/antarctic_plots#215] [Q&A] [answered]
Hi, community. I am using Polartoolkit to generate the bathymetry under Ross Ice Shelf. It is better if I can shade the Ross Ice Shelf region with some transparent color. However, I can only do "shading" command under the "maps.plot_grd", and the shading will not show the Bathymetry details under the Ice Shelf. May I know if anyone knows how to apply a transparent shading for ice shelves? Thanks
![RIS_bathymetry_shading](https://github.com/mdtanker/polartoolkit/assets/136052950/3a0ee294-861d-4bf7-ba44-a1f43047a41c)

**Comment by mdtanker** [comment]
Hi @YingpuRin thanks for opening a discussion! Do you mind sharing the code you used to create the figure? 

So you want the map to be colored by  bathymetry under the ice shelf. What do you want to be transparent? The ice surface elevation? 

I'm happy to help out, but I'm just a little confused what you're trying to do. 

Below are a few example Ross Ice Shelf figures I've made before in case they give you any ideas for what you want.

https://github.com/mdtanker/phdthesis/blob/main/figures%2Fchp2%2FFig1_Bathy_Mag.jpg
   
https://github.com/mdtanker/phdthesis/blob/main/figures%2Fchp4%2FRIS_inversion_results.png

https://github.com/mdtanker/phdthesis/blob/main/figures%2Fchp2%2Ffigure_S1.png

**Comment by YingpuRin** [comment]
Hi Matt,

Thanks for your reply. Btw, we actually met before at Vic Uni 🙂 I am Yingpu from the ASP NZ.

So, my colleague created a map using quantarctica, and I want to do something similar with PolarToolkit. If you can see the attachment, my colleague put a shading on top of the Ross Ice Shelf. As we are doing oceanography study, we focus on the bathymetry under the Ross Ice Shelf, and we want a gray shading to illustrate the location of the Ross Ice Shelf.

I also attached my code in the email, and it seems like the "transparency"  command does not work under "maps.plot_grd" from my side.

Looking forward to your reply,
Yingpu
________________________________
发件人: Matt Tankersley ***@***.***>
发送时间: 2024年6月28日 18:38
收件人: mdtanker/polartoolkit ***@***.***>
抄送: YingpuRin ***@***.***>; Mention ***@***.***>
主题: Re: [mdtanker/polartoolkit] Shading Ice Shelf with Polartoolkit (Discussion #215)


Hi @YingpuRin<https://github.com/YingpuRin> thanks for opening a discussion! Do you mind sharing the code you used to create the figure?

So you want the map to be colored by bathymetry under the ice shelf. What do you want to be transparent? The ice surface elevation?

I'm happy to help out, but I'm just a little confused what you're trying to do.

Below are a few example Ross Ice Shelf figures I've made before in case they give you any ideas for what you want.

https://github.com/mdtanker/phdthesis/blob/main/figures%2Fchp2%2FFig1_Bathy_Mag.jpg

https://github.com/mdtanker/phdthesis/blob/main/figures%2Fchp4%2FRIS_inversion_results.png

https://github.com/mdtanker/phdthesis/blob/main/figures%2Fchp2%2Ffigure_S1.png

**Reply by mdtanker** [reply]
Yes I thought I recognized your name 🙂 Sorry there was no attachment, can you try and include it another way?

**Reply by YingpuRin** [reply]
Sorry. I was replying in emails. Now I attached the fig and my code here on Github. It should work this time. 

Thanks so much for your support. 

Yingpu


![Craig_Quantarctica](https://github.com/mdtanker/polartoolkit/assets/136052950/d2148a56-a779-4bec-9ec7-883728cb75ab)



Below is my Python code:

```python
#https://polartoolkit.readthedocs.io/en/latest/index.html
from polartoolkit import fetch, maps, profiles, regions, utils
import os
import pandas as pd
import numpy as np
import pyogrio
import pygmt

os.environ["POLARTOOLKIT_HEMISPHERE"] = "south"

# manually define a geographic region in meters in the South Polar Stereograph
# Projection (EPSG) in the format (x_min, xmax, y_min, y_max)
#EPSG 3031 projection

#defined_region = (-13190874, 13164423, -71859306, -71021912)


defined_region = regions.ross_ice_shelf 

#Using Bedmap2 data for bathymetry
#https://tc.copernicus.org/articles/7/375/2013/
#layer=="bed": Bedmap2 bed elevation grid
# bedmap2_bed 				is bed height
#bedmap2_icemask_grounded_and_shelves is a mask file showing the grounding line and the extent of the floating ice shelves
#https://secure.antarctica.ac.uk/data/bedmap2/bedmap2_readme.txt
#https://www.bas.ac.uk/project/bedmap-2/#data
water_thickness = fetch.bedmap2(
    layer="bed",
    #layer="water_thickness",
    #region=regions.ross_ice_shelf,
    region=defined_region,
)

icemask_grounded_and_shelves= fetch.bedmap2(
    layer="icemask_grounded_and_shelves",
    #layer="water_thickness",
    #region=regions.ross_ice_shelf,
    region=defined_region,
)



#Masking the outside RIS region
#https://antarctic-plots.readthedocs.io/en/latest/gallery/mask_from_shp.html
# fetch a shapefile
#The locations of the groundingline and coastline for Antarctica from various sources.
#Depoorter, Mathieu A., Jonathan L. Bamber, Jennifer Griggs, Jan T. M. Lenaerts, Stefan R. M. Ligtenberg, Michiel R. van den Broeke, and Geir Moholdt. “Antarctic Masks (Ice-Shelves, Ice-Sheet, and Islands), Link to Shape File.” In Supplement to: Depoorter, MA et al. (2013): Calving Fluxes and Basal Melt Rates of Antarctic Ice Shelves. Nature, 502, 89-92, Https://Doi.Org/10.1038/Nature12567. PANGAEA, 2013. https://doi.org/10.1594/PANGAEA.819147     
#https://polartoolkit.readthedocs.io/en/latest/datasets/antarctica/groundingline.html
shapefile = fetch.groundingline(version="depoorter-2013")
# get just the grounding line
gdf = pyogrio.read_dataframe(shapefile)
groundingline = gdf[gdf.Id_text == "Grounded ice or land"]

# mask the inside region
masked_inside = utils.mask_from_shp(shapefile=groundingline, xr_grid=water_thickness, masked=True)

# not working for plotting
#masked_water_thickness = water_thickness.to_masked_array((a ==0) for a in icemask_grounded_and_shelves)


#Get the shape of ice shelf
ice_shelf_Ross = gdf[gdf.Id_text == "Ice shelf"]



fig = maps.plot_grd(
    masked_inside,
    #region=regions.ross_ice_shelf,
    #title="Water thickness of the Ross Ice Shelf",
    cbar_label="meters",
    cmap="globe",
    #cmap="lajolla",
    #cmap="vik",
    #cmap="balance+h0",
    #cmap="vik",
    inset=True,
    coast=True,
    scalebar=True,
    gridlines=True,
    x_spacing= 20,
    north_arrow=True,
    #robust=True,
    #hist=True,
    #shading command
    #https://polartoolkit.readthedocs.io/en/latest/_modules/polartoolkit/maps.html#
    #shading=(icemask_grounded_and_shelves, transparency==0.2), 
    #shading="+a45+nt1",   
    #transparency=50,
)


fig.plot(ice_shelf_Ross, pen="1p,black")



#maps.basemap(
#    region=regions.ross_ice_shelf,
#    hemisphere="south",
#    origin_shift = "initialize",
#    points=icemask_grounded_and_shelves,
#    #fetch.bedmap2(layer="icemask_grounded_and_shelves"),
#)


#fig.savefig('RIS_WaterColumn_masked.png')
fig.savefig('RIS_bathymetry_shading.png')
#fig.savefig('RIS_bathymetry_masked.png')
```

**Reply by mdtanker** [reply, marked as answer]
Thanks for providing those. So the parameter `shading` is used to provide shadows to the grid, while `transparency` can be used to make the grids transparent. Below is some code and figure that should give you a better starting point! Please let me know if you need any more help :) 
PyGMT will have instructions for adding the text, labelled locations, and contours to your map.

![image](https://github.com/mdtanker/polartoolkit/assets/81199856/ecbfc978-dd1e-436c-8904-edf31c3d70dd)


```python
from polartoolkit import fetch, maps, regions
import os
import numpy as np
import pygmt
import xarray as xr

os.environ["POLARTOOLKIT_HEMISPHERE"] = "south"

# region = regions.ross_ice_shelf
region = (-680000.0, 470000.0, -1620000.0, -310000.0)

bed = fetch.bedmap2(
    layer="bed",
    region=region,
)

# grid with 1 for ice shelves, 0 for grounded ice and NaNs for open ocean
mask = fetch.bedmap2(
    layer="icemask_grounded_and_shelves",
    region=region,
)
ice_shelf = xr.where(mask==1, bed, np.nan)
offshore = xr.where(mask.isnull(), bed, np.nan)

# plot a basemap with gray background
fig = maps.basemap(
    inset=True,
    fig_height=20,
    region=region,
    frame="+ggray",
)

# make a cmap with the entire grid
pygmt.grd2cpt(
    bed,
    cmap="ice",
    limit=[-3000,0],
    continuous=True,
    background=True,
    output="bed.cpt",
)

# plot the offshore data
fig = maps.plot_grd(
    offshore,
    fig=fig,
    fig_height=20,
    origin_shift="no_shift",
    cbar_label="meters",
    cmap="bed.cpt",
    colorbar=False,
)

# plot transparent ice shelf data
fig = maps.plot_grd(
    ice_shelf,
    fig=fig,
    fig_height=20,
    origin_shift="no_shift",
    cbar_label="meters",
    cmap="bed.cpt",
    inset=True,
    coast=True,
    scalebar=True,
    gridlines=True,
    x_spacing= 20,
    north_arrow=True,
    transparency=50,
)

fig.show()
```

**Reply by YingpuRin** [reply]
Hi Matt. Thanks for your reply. I found out that my transparent command does not work as there is a compatible issue with my setup. I fixed the problem, and now the transparency for ice shelves works for me. Thanks so much for sharing the codes, and for sure I learnt a lot from your code. Now, I am improving my codes, and your plotting method with " xr.where" function helps me a lot in doing this! Thanks a lot for your help, and I cannot know this method without your sharing.
